DAOS-17427 control: Restart excluded rank after suicide

tanabarr opened the pull request:
When an engine detects that it has been removed from the system group
map by receiving a CART event, it will now notify its local control plane
with a RAS engine_self_terminated event before terminating its own
process. After receiving this self-termination event, the local
control plane will restart the engine so it can rejoin the system.

The goal of this change is to improve overall system resilience by
automatically recovering engines that are excluded because of
temporary issues such as network instability. Once the engines rejoin,
the rank will still need to be reintegrated into pools as a separate
follow‑up step.

Rate-limiting prevents restart storms: a configurable minimum delay
(default 300 seconds) between restarts per rank ensures system
stability. Two new server config file parameters control behavior:
disable_engine_auto_restart (boolean, default false) completely
disables automatic restarts, while engine_auto_restart_min_delay
(integer seconds) sets the minimum time between consecutive restart
attempts.

Functional tests for the automatic engine restart feature included
with cases to verify disabling, rate-limiting and configuration support.

Features: pool control

### Steps for the author:

* [x] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [x] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [x] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [x] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [x] Gatekeeper requested (daos-gatekeeper added as a reviewer).

tanabarr committed:
DAOS-17427 control: Restart evicted rank after suicide

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-16279/1/execution/node/1466/log

tanabarr committed:
implement suicide event handlers

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

liw reviewed (commented):
Not requesting changes.

tanabarr committed:
add unit testing and documentation

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

kjacque commented on src/engine/init.c:
```diff
@@ -597,17 +597,21 @@ dss_crt_event_cb(d_rank_t rank, uint64_t incarnation, enum crt_event_source src,
+				D_ERROR("failed to handle %u/%u event: " DF_RC "\n", src, type,
+					DP_RC(rc));
+
 			rc = kill(getpid(), SIGKILL);
```
I agree we should start conservatively. For engines that crash, or exit for some unknown reason, we don't know if the engine would be OK if we restarted it. Better to let the admin investigate and manually restart the ranks in that case.

tanabarr committed:
fix docs and unit tests

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
revise unit test for suicide handler

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
fixup tests

Features: control pool

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/5/execution/node/1303/log

tanabarr committed:
Merge branch 'master' into tanabarr/control-engine-suicide-restart

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

tanabarr committed:
rename suicide to self terminated

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr commented on src/control/server/server_utils.go:
```diff
@@ -777,6 +815,15 @@ func registerFollowerSubscriptions(srv *server) {
 	srv.pubSub.Reset()
 	srv.pubSub.Subscribe(events.RASTypeAny, srv.evtLogger)
 	srv.pubSub.Subscribe(events.RASTypeStateChange, srv.evtForwarder)
+	srv.pubSub.Subscribe(events.RASTypeInfoOnly,
```
That isn't strictly accurate because in the function we register a subscription for an event forwarder which is not "LocalOnly" I will go with `registerNonLeaderSubscriptions` instead.

tanabarr committed:
rename registerFollowerSubscriptions to registerSubscriptions

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
add flag to disable automatic engine restart

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
fix intermittent test fails with delay before txt comp

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Merge branch 'tanabarr/control-engine-suicide-restart' of github.com:daos-stack/daos into tanabarr/control-engine-suicide-restart

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
implement basic rate limiting

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
improve naming consistency and fix config unit tests

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
add rate-limiting unit test

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr commented on docs/overview/fault.md:
```diff
@@ -84,3 +84,28 @@ can now read from the rebuilt object shards.
+
+A DAOS engine may be excluded from the group map because of inactivity
+for example. When an engine becomes aware of it's removal from the
+group map it will self-terminate (referred to as "suicide") to protect
```
done

tanabarr committed:
documentation updates

Doc-only: true
Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

kjacque commented on src/control/server/config/server.go:
```diff
@@ -1,6 +1,6 @@
-// (C) Copyright 2020-2024 Intel Corporation.
-// (C) Copyright 2025-2026 Hewlett Packard Enterprise Development LP
+// Copyright 2020-2024 Intel Corporation.
+// Copyright 2025-2026 Hewlett Packard Enterprise Development LP
```
Probably going to have to revert this. I think the commit that updated the copyright githook was reverted.

tanabarr replied:
thanks for the heads up

kjacque reviewed (changes requested):
Thanks for cleaning up the terminology.

kjacque commented on src/control/lib/control/event.go:
```diff
@@ -170,7 +171,8 @@ func newEventLogger(logBasic logging.Logger, newSyslogger newSysloggerFn) *Event
 // NewEventLogger returns an initialized EventLogger capable of writing to the
-// supplied logger in addition to syslog.
+// supplied logger in addition to syslog. Should only be used in production code,
+// use MockEventLogger in unit tests.
```
Okay, that's fair.

tanabarr commented on docs/overview/fault.md:
```diff
@@ -89,23 +89,58 @@ and updating objects.
+are rate-limited on a per-rank basis. By default, a minimum delay of 300 seconds
+(5 minutes) is enforced between consecutive restart attempts for the same rank.
+If an engine self-terminates again before this delay expires, the restart request
+is rejected and logged at NOTICE level.
```
not currently, it will just wait for the next RAS

tanabarr commented on src/control/go.mod:
```diff
@@ -5,7 +5,6 @@ module github.com/daos-stack/daos/src/control
 // - debian packaging version checks: debian/control
 // Scons uses this file to extract the minimum version.
 go 1.21
-toolchain go1.23.0
```
done

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart
fix copyright and logging levels
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Q a single restart request if received within timeout period

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16279/13/display/redirect

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/14/execution/node/421/log

tanabarr commented on src/control/server/server_utils_test.go:
```diff
@@ -2206,7 +2228,131 @@ func TestServer_handleEngineSuicide(t *testing.T) {
+	// Wait for the EngineAutoRestartMinDelay to expire
+	time.Sleep(time.Duration(srv.cfg.EngineAutoRestartMinDelay) * time.Second)
+
+	// Third restart after delay should succeed
```
as discussed, rate-limiting has been implemented so a new event won't be added

kjacque commented on src/control/server/server_utils_test.go:
```diff
@@ -2312,22 +2314,44 @@ func TestServer_handleEngineSelfTerminated_RateLimiting(t *testing.T) {
 	}
 
-	// Wait for the EngineAutoRestartMinDelay to expire
+	// Verify pending timer exists
+	srv.rankRestartMu.Lock()
+	if _, exists := srv.rankRestartPending[evt.Rank]; !exists {
+		srv.rankRestartMu.Unlock()
+		t.Fatal("expected pending restart timer to exist")
+	}
+	srv.rankRestartMu.Unlock()
```
minor - might be better to do this in a function/closure so you can `defer srv.rankRestartMu.Unlock()` instead.

tanabarr replied:
done

mjmac commented on src/control/server/config/server.go:
```diff
@@ -98,7 +98,9 @@ type Server struct {
-	AutoFormat bool `yaml:"-"`
+	AutoFormat                bool `yaml:"-"`
+	DisableEngineAutoRestart  bool `yaml:"disable_engine_auto_restart"`
+	EngineAutoRestartMinDelay int  `yaml:"engine_auto_restart_min_delay,omitempty"`
```
There doesn't appear to be any validation of this parameter, e.g. in the `Validate()` method:

```go
       if cfg.EngineAutoRestartMinDelay < 0 {
               return errors.Errorf("engine_auto_restart_min_delay must be >= 0 (got %d)",
                       cfg.EngineAutoRestartMinDelay)
       }
```

tanabarr replied:
done

mjmac commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
-func registerFollowerSubscriptions(srv *server) {
+// Handle local engine self termination and restart engine to rejoin system.
+func handleEngineSelfTerminated(ctx context.Context, srv *server, evt *events.RASEvent) error {
+
```
Maybe add a check to guard against processing a non-local event that got here somehow?

```suggestion
       if evt.IsForwarded() || (evt.Hostname != "" && evt.Hostname != srv.hostname) {
               srv.log.Debugf("ignoring non-local engine_self_terminated event from %q", evt.Hostname)
               return
       }
```

tanabarr replied:
happy to add but should not be needed because these messages are not forwarded

tanabarr replied:
done

kjacque commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
+	if len(instances) == 0 {
+		return errors.Errorf("no instance found for rank %d", evt.Rank)
+	}
+	if len(instances) > 1 {
+		return errors.Errorf("multiple instances found for rank %d", evt.Rank)
+	}
```
Might be inclined not to check this one -- or just log an error instead of failing hard here.

tanabarr replied:
why would you ever want to have multiple instances returned for the same rank on a host? surely something is seriously broken in that case

kjacque replied:
Yeah it would definitely be a sign of something very strange, but I'm wondering how we'd get into such a situation to begin with. Doesn't hurt to catch it I guess. I know sometimes we can be a little too stringent about vetting what's returned, making it so the user is basically stuck with no information, but that's more relevant to dmg and the daos tool.

kjacque commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
+						evt.Rank, engine.Index())
+					srv.rankRestartMu.Lock()
+					delete(srv.rankRestartPending, evt.Rank)
+					srv.rankRestartMu.Unlock()
```
I wonder if it would be better to delegate the stuff that needs to be wrapped in the lock to a helper function. Then you could use `defer srv.rankRestartMu.Unlock()` within that scope, and not have to worry about unlocking before the various returns here.

tanabarr replied:
done

kjacque reviewed (dismissed):
Comments are minor and not something I'd block on. I do think using `defer` makes it a lot easier to be sure we're handling mutexes safely.

tanabarr commented:
@kjacque I will address the nonblocking comments if I have to repush. @mjmac would you be able to review as I can't seem to find anyone else comfortable with reviewing? TIA

mjmac commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
+	if len(instances) > 1 {
+		return errors.Errorf("multiple instances found for rank %d", evt.Rank)
+	}
+	engine := instances[0]
```
nit: The variable `engine` shadows the imported package name. Might be safer to use `ei` like other areas of the code. I seem to recall that we had subtle problems caused by this in the past.

tanabarr replied:
done

mjmac commented on src/control/server/server.go:
```diff
@@ -167,6 +167,10 @@ type server struct {
 	onShutdown       []func()
+
+	rankRestartMu      sync.Mutex
+	rankRestartTimes   map[uint32]time.Time
+	rankRestartPending map[uint32]*time.Timer
```
nit: Hmm. I know that `Event.Rank` is `uint32`, but maybe it should have been `ranklist.Rank`? We've been down this road before, where we used bare `uint32` to represent a rank because otherwise it meant type conversion, but IMO short-term convenience leads to longer-term decreases in code readability and quality. My preference would be to avoid `uint32` for rank in the control plane as much as possible.

tanabarr replied:
done

mjmac commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
 	srv.pubSub.Subscribe(events.RASTypeAny, srv.evtLogger)
 	srv.pubSub.Subscribe(events.RASTypeStateChange, srv.evtForwarder)
+	srv.pubSub.Subscribe(events.RASTypeInfoOnly,
+		events.HandlerFunc(func(ctx context.Context, evt *events.RASEvent) {
```
Maybe this should be extracted into a helper used for both leader/follower paths? Seems like the code is identical in both places.

tanabarr replied:
done

mjmac commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +774,126 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
-// This is the initial behavior before leadership has been determined.
-func registerFollowerSubscriptions(srv *server) {
+// Handle local engine self termination and restart engine to rejoin system.
+func handleEngineSelfTerminated(ctx context.Context, srv *server, evt *events.RASEvent) error {
```
This function is really doing a lot, and it's quite difficult to reason about its behavior over time and in a concurrent environment. I strongly suspect that there is at least once race involving execution of the callbacks after the timer. The lock protects against multiple invocations of the handler, but I don't think it correctly protects against a handler making a decision based on a timestamp that's about to be updated by a callback (because the lock is only taken in the callback when it updates the timestamp).

I'm not 100% confident that it's wrong, but I'm also really not confident that it's right. I think it would be a lot cleaner to use a lockless approach and instead have the restart logic sitting out in its own thing that listens on a channel. The main logic can make restart requests all day long, but the restart manager only queues a new request if there isn't one already in progress. Aside from being easier to read and reason about, formalizing the restart logic as a concrete unit of code should also make it much easier to test thoroughly.

kjacque replied:
This is a good idea.

tanabarr replied:
I will do that as a separate Commit from the rest of the review comment changes

tanabarr replied:
done

tanabarr committed:
address review comments from mjmac and kjacque

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
use channel-based restart manager for rate-limiting

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
fix handleEngineSelfTerminated unit tests

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
add unit tests for engine restart manager

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart
Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
remove deprecated code

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart
Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/19/testReport/

tanabarr committed:
DRY-up unit tests for engine restart manager

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/20/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/22/testReport/

tanabarr commented:
CI failing on one dfuse known issue (seeing on other PRs)

tanabarr commented:
@mjmac @kjacque please could I have rereviews, I've refactored as per the review comments

tanabarr committed:
Merge branch 'master' into tanabarr/control-engine-suicide-restart

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

tanabarr commented:
@mjmac @kjacque ping for reviews please

daosbuild3 commented:
Test stage Unit Test completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/23/testReport/

tanabarr committed:
DAOS-17427 test: Auto-restart after self-terminate tests (#18006)

Functional tests for the automatic engine restart feature introduced
in the control plane. These tests verify that engines automatically
restart after self-termination when excluded from the system, with
cases to verify disabling, rate-limiting and configuration support.

Test-tag: hw,medium,dmg,control,engine_auto_restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

* try to fix test issues

Test-tag: hw,medium,dmg,control,engine_auto_restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Unit Test completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/25/testReport/

kjacque commented on src/control/server/ctl_ranks_rpc.go:
```diff
@@ -206,6 +206,22 @@ func (svc *ControlService) StopRanks(ctx context.Context, req *ctlpb.RanksReq) (
+		if len(ranks) > 0 {
+			svc.restartMgr.clearRankRestartHistory(ranks)
+		}
+	}
```
minor - I'd be inclined to break this out into a helper function to reduce duplication. Would also make this logic easy to test.

tanabarr replied:
done

kjacque commented on src/control/server/server.go:
```diff
@@ -167,6 +168,13 @@ type server struct {
+	// Deprecated: use restartMgr instead
+	rankRestartMu      sync.Mutex
+	rankRestartTimes   map[ranklist.Rank]time.Time
+	rankRestartPending map[ranklist.Rank]*time.Timer
```
Shouldn't this be removed from the PR?

tanabarr replied:
done

kjacque commented on src/control/server/server.go:
```diff
@@ -283,6 +295,11 @@ func (srv *server) OnShutdown(fns ...func()) {
 
 func (srv *server) shutdown() {
+	// Stop the restart manager first
+	if srv.restartMgr != nil {
```
Is it allowed to be nil? Seems better to initialize it in the constructor, and then assume that it's initialized correctly so you don't have to do all these nil checks.

tanabarr replied:
done

kjacque commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +769,82 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
+	if srv.restartMgr == nil {
+		return errors.Errorf("restart manager not initialized, cannot restart rank %d",
+			rank)
+	}
```
Same question here about the nil checks... Maybe restartMgr should check if it's nil and return errors from its methods if so.

tanabarr replied:
removed the nil checks and will rely on constructor

kjacque commented on src/control/server/server_utils.go:
```diff
@@ -769,14 +769,82 @@ func registerTelemetryCallbacks(ctx context.Context, srv *server) {
+func subscribeEngineSelfTerminated(srv *server) events.Handler {
+	return events.HandlerFunc(func(ctx context.Context, evt *events.RASEvent) {
+		switch evt.ID {
+		case events.RASEngineSelfTerminated:
```
Any reason to use a `switch` here if there's only one allowed type?

tanabarr replied:
done

kjacque commented on src/control/server/instance_restart.go:
```diff
@@ -0,0 +1,246 @@
+
+	// Record restart time and clear pending state on exit (deferred)
+	mgr.recordRestartTime(rank)
+	mgr.log.Noticef("recording rank %d", rank)
```
Maybe shouldn't be Notice for this one. Seems like this is mostly useful for debug.

tanabarr replied:
done

kjacque replied:
Oops, looks like this fix didn't get pushed.

tanabarr replied:
got inadvertently reverted in bd15522c203539d71553a25b25c6684f10f95067 , added back, done

kjacque commented on src/control/server/instance_restart.go:
```diff
@@ -0,0 +1,246 @@
+	}
+	mgr.pendingRestart = make(map[ranklist.Rank]*time.Timer)
+
+	close(mgr.stopChan)
```
`start` should re-create the channel. Otherwise `stop` won't work again after closing it.

tanabarr replied:
the design is to only start once on service start and stop on  shutdown so this isn't necessary for practical purposes, I will add the capability anyway as it's nice to be re-entrant

kjacque replied:
As discussed, I agree it's fine as-is. I do think it'd be nice to leave a note to ourselves as a comment though.

tanabarr replied:
comments added

kjacque commented on src/control/server/instance_restart_test.go:
```diff
@@ -0,0 +1,748 @@
+
+// Test helper functions
+
+func setupTestLogger(t *testing.T) (logging.Logger, *logging.LogBuffer) {
+	t.Helper()
+	log, buf := logging.NewTestLogger(t.Name())
+	t.Cleanup(func() {
+		test.ShowBufferOnFailure(t, buf)
+	})
+
+	return log, buf
+}
```
Don't want to use `test.MustLogContext`/`logging.FromContext` to create the logger?

tanabarr replied:
the problem is that this Pattern doesn't enable me to expose the LogBuffer to be used in the tests for comparison purposes. maybe I'm missing a trick here? @mjmac

kjacque replied:
Ah, I see! Never mind, I'm not sure how to extract the buffer in that case either. Might be nice to add this function to the `test` package. I suspect this will be useful in other tests (maybe call it `NewTestLoggerWithBuffer` or something like that?). Not requesting a change on this PR, just a suggestion.

kjacque commented on src/control/server/instance_restart_test.go:
```diff
@@ -0,0 +1,748 @@
+	}
+
+	// Wait for timer to fire (with buffer)
+	time.Sleep(10 * time.Second)
```
Probably don't need to wait this long if the timer is firing around 2 seconds. Maybe 3-5 sec is more reasonable?

kjacque commented on src/control/server/instance_restart_test.go:
```diff
@@ -0,0 +1,748 @@
+	}
+
+	// Cleanup
+	if timer != nil {
+		timer.Stop()
+	}
```
For timers that need to stop at the end, you could use `defer` after you get them.

tanabarr replied:
in this case we don't know whether the timer still exists at end of func, could do the nil check in defer func but I don't think that buys us much

kjacque replied:
Looks like the test asserts if the timer wasn't set in the map... So it would have to be non-nil at this point, wouldn't it? Doesn't look like we're ever resetting that variable.

tanabarr replied:
have changed to improve clarity, done

kjacque commented on src/control/server/instance_restart_test.go:
```diff
@@ -0,0 +1,748 @@
+	cfg := &config.Server{
+		EngineAutoRestartMinDelay: 42,
+	}
+	mgr := setupTestManager(t, cfg)
```
Should probably call `NewEngineRestartManager` directly here to verify only that behavior, not the test constructor.

tanabarr replied:
done

kjacque commented on src/control/server/server_utils_test.go:
```diff
@@ -1981,3 +1988,753 @@ f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
+	}
+}
+
+func TestServer_handleEngineSelfTerminated_RateLimiting(t *testing.T) {
```
Since the detailed behavior is now covered in the engineRestartManager, you can remove some of these higher level tests. You could just make sure the request is sent to the restart manager.

tanabarr replied:
I'm going to leave them in for the moment unless you feel particularly strongly

kjacque replied:
I don't feel that strongly, fine with leaving them.

kjacque reviewed (changes requested):
First of all: nice work, and excellent job with all the test coverage. My comments are mostly minor cleanup stuff, but the stop/start behavior is something we may want to make behave better, or at least note that we only get to stop it once.

tanabarr committed:
fix server package unit test helpers

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/25/execution/node/983/log

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Revert "fix server package unit test helpers"

This reverts commit bdfde05ff85538e54984d163353429dfb9ff52f4.

tanabarr committed:
fix server package unit test helpers

Features: control
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
addressed review comments from kjacque pt1

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
allow restart manager to close and open again

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Revert "allow restart manager to close and open again"

This reverts commit 049509cedbd3ebfa67b0bd4abc06a7fc3898abb0.

tanabarr committed:
address some review comments from kjacque

Features: control full_regression
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

kjacque commented on src/control/server/ctl_ranks_rpc.go:
```diff
@@ -206,21 +221,9 @@ func (svc *ControlService) StopRanks(ctx context.Context, req *ctlpb.RanksReq) (
-			svc.restartMgr.clearRankRestartHistory(ranks)
-		}
-	}
+	// clearly state history for stopped ranks, instances have already been filtered by
```
typo? clearly -> clear

tanabarr replied:
done

kjacque commented on src/control/server/ctl_ranks_rpc.go:
```diff
@@ -335,21 +338,9 @@ func (svc *ControlService) StartRanks(ctx context.Context, req *ctlpb.RanksReq)
-			svc.restartMgr.clearRankRestartHistory(ranks)
-		}
-	}
+	// clearly state history for started ranks, instances have already been filtered by
```
typo? clearly -> clear

tanabarr replied:
done

kjacque commented on src/control/server/ctl_ranks_rpc_test.go:
```diff
@@ -76,17 +76,15 @@ func checkUnorderedRankResults(t *testing.T, expResults, gotResults []*sharedpb.
+		ei.ready.SetTrue()
 	}
-	srv.runner = engine.NewTestRunner(trc, engine.MockConfig())
-	srv.setIndex(idx)
```
I guess no tests were using the index?

tanabarr replied:
nope, so I removed it

kjacque reviewed (dismissed):
Nothing left that's blocking, just the minor typos and the Notice level logging should be fixed. I'm fine with that happening in a follow-on.

daosbuild3 commented:
Test stage Functional on EL 9 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/26/testReport/

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
comment one start/stop per process lifetime

Test-tag: pr control full_regression
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+            self.log.error("Failed to reset engine restart state: %s", error)
+            self.fail("tearDown failed to reset engine restart state: {}".format(error))
+        finally:
+            super().tearDown()
```
We have a way to handle this in the framework by calling
```
self.register_cleanup(reset_engine_restart_state)
```
After whatever operation puts the system into the invalid state. So maybe instead of defining this tearDown, we can do that after calling `exclude_rank_and_wait_restart` the first time

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+            2. Wait for rank to self-terminate
+            3. Verify rank automatically restarts and rejoins the system
+
+        :avocado: tags=all,pr,daily_regression
```
Just confirming that we really do want to add this to PR testing?

tanabarr replied:
So yes I think so it's a significant feature

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        all_ranks = self.get_all_ranks()
+        if len(all_ranks) < 2:
+            self.skipTest("Test requires at least 2 ranks")
```
It is better to fail because skipping will be silent and easily ignored
```suggestion
            self.fail("Test requires at least 2 ranks")
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        self.log.info("rank %s initial incarnation: %s", test_rank, initial_incarnation)
+
+        restarted, final_state = self.exclude_rank_and_wait_restart(test_rank)
+
+        if not restarted:
```
nit - intentional new line?
```suggestion
        restarted, final_state = self.exclude_rank_and_wait_restart(test_rank)
        if not restarted:
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        # verify incarnation increased after restart
+        final_incarnation = self.get_rank_incarnation(test_rank)
+        if final_incarnation is None:
+            self.fail(f"failed to get final incarnation for rank {test_rank}")
```
It would be better if get_rank_incarnation raised an exception instead of silently returning None

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        all_ranks = self.get_all_ranks()
+        if len(all_ranks) < 3:
+            self.skipTest("Test requires at least 3 ranks")
```
```suggestion
            self.fail("Test requires at least 3 ranks")
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        num_to_test = max(2, len(all_ranks) // 2)
+        test_ranks = self.random.sample(all_ranks, num_to_test)
+
+        self.log_step("Step 1: Excluding %s ranks: %s", (num_to_test, test_ranks))
```
IMO it's better to not put `Step` in the string when using `log_step` because `log_step` already adds step numbers. Also, `log_step` does not support the `%s` syntax, which would raise an exception if this test was running in CI
```suggestion
        self.log_step(f"Exclude {num_to_test} ranks: {test_ranks}")
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+            if initial_incarnation is None:
+                self.fail(f"failed to get initial incarnation for rank {rank}")
+            incs.append(initial_incarnation)
+            self.dmg.system_exclude(ranks=[rank], rank_hosts=None)
+            time.sleep(1)  # small delay between exclusions
+            self.dmg.system_clear_exclude(ranks=[rank], rank_hosts=None)
```
Where does the arbitrary 1s sleep come from? Historically, this kind of thing eventually fails and needs to be adjusted

tanabarr replied:
it's to enable the propagation of exclude state within the control-plane before we clear the exclusion, closes a potential race where state is cleared before propagated which would prevent self-determination of engine process.

daltonbohning replied:
How do we know 1s is enough?

tanabarr committed:
address more review comments from kjacque

Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        end_incs = []
+        for rank in test_ranks:
+            failed = self.server_managers[0].check_rank_state(
+                ranks=[rank], valid_states=["joined"], max_checks=1)
```
Instead of arbitrarily sleeping for 35 seconds, why not increase `check_rank_state(..., max_checks=1)` to something higher?

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+        for idx, (old, new) in enumerate(zip(incs, end_incs)):
+            actual_rank = test_ranks[idx]
+            if new > old:
+                self.log.debug(f"Rank {actual_rank}: {old} -> {new} (restarted)")
```
> logging-fstring-interpolation, Use lazy % formatting in logging functions

tanabarr replied:
fixed

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+            self.skipTest("Test requires at least 4 ranks")
+
+        # Create pool first
+        self.add_pool(connect=False)
```
We're trying to not use that function for new tests. Instead:

```suggestion
        pool = self.get_pool(connect=False)
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.py:
```diff
@@ -0,0 +1,201 @@
+
+        # Verify pool is still accessible
+        self.log_step("Verifying pool is still accessible after rank restart")
+        self.pool.query()
```
And then use the local `pool` variable here
```suggestion
        pool.query()
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.yaml:
```diff
@@ -0,0 +1,25 @@
+hosts:
+  test_servers: 2
+server_config:
```
We typically put the timeout up here
```suggestion
hosts:
  test_servers: 2
timeout: 300
server_config:
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.yaml:
```diff
@@ -0,0 +1,25 @@
+          scm_mount: /mnt/daos1
+pool:
+  size: 2G
+timeout: 300
```
```suggestion
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +48,176 @@ def verify_dmg_storage_scan(self, verify_method):
+        """
+        data = self.dmg.system_query(ranks="%s" % rank)
+        if data["status"] != 0:
+            self.fail("Cmd dmg system query failed")
```
It's better for a helper function to raise an exception and let the test itself decide if it is a failure
```suggestion
        if data["status"] != 0:
            raise CommandFailure("dmg system query failed")
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +48,176 @@ def verify_dmg_storage_scan(self, verify_method):
+            data = self.dmg.system_query(ranks=f"{rank}")
+            if data.get("status") != 0:
+                self.log.error("dmg system query failed for rank %s", rank)
+                return None
```
IMO it's better to raise exception than silently skip
```suggestion
            if data.get("status") != 0:
                raise CommandFailure(f"dmg system query failed for rank {rank}")
```

tanabarr replied:
done

daltonbohning commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +48,176 @@ def verify_dmg_storage_scan(self, verify_method):
+        except Exception as error:  # pylint: disable=broad-exception-caught
+            # Catch all exceptions to prevent test framework crashes during rank queries
+            self.log.error("Exception getting incarnation for rank %s: %s", rank, error)
+            return None
```
This comment is indicative of a larger problem. An unhandled exception should, IMO, cause the test to ultimately fail. We should not silently skip something that is broken

tanabarr replied:
fixed

daltonbohning commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +48,176 @@ def verify_dmg_storage_scan(self, verify_method):
+        self.log.info("Restarting servers to reset engine restart manager state")
+        self.server_managers[0].system_stop()
+        time.sleep(2)
+        self.server_managers[0].system_start()
```
Where does the arbitrary 2s sleep come from? This will eventually be a problem and we will have to revisit

tanabarr replied:
will remove if you don't think it necessary for a system restart

tanabarr replied:
done

daltonbohning replied:
If we don't need it at all then that is great, but I was more meaning that 2s seems arbitrary. And historically arbitrary sleeps have been an issue that we have to revisit later

daltonbohning reviewed (commented):
It seems the new functional tests aren't even running in CI?
There are many lint errors in ftest that need to be resolved or possibly ignored

tanabarr committed:
pylint fixes

Test-tag: pr control full_regression
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/28/execution/node/942/log

tanabarr committed:
using self.register_cleanup (#18240)

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>
Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>

tanabarr committed:
Apply suggestion from @daltonbohning

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
more ftest related review comment updates

Test-tag: hw,medium,dmg,control,engine_auto_restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +48,170 @@ def verify_dmg_storage_scan(self, verify_method):
+
+        Returns:
+            tuple: (restarted, final_state) - whether rank restarted and its final state
+        """
```
moving register_cleanup() to beginning of each test because not all of them call this helper

tanabarr committed:
f-string updates and remove step comments in log_step calls use CommandFailure exception for helpers and register cleanup in setUp for each test class

Test-tag: hw,medium,dmg,control,engine_auto_restart
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Merge remote-tracking branch 'origin/master' into tanabarr/control-engine-suicide-restart

Test-tag: hw,medium,dmg,control,engine_auto_restart pr
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.yaml:
```diff
@@ -0,0 +1,25 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos0
```
We should use `storage: auto` for HW tests so the framework can set this up for both pmem and md on ssd
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart.yaml:
```diff
@@ -0,0 +1,25 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos1
```
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_advanced.py:
```diff
@@ -0,0 +1,156 @@
+
+        if actual_delay < min_delay:
+            self.fail(f"Restart too early: {actual_delay:.1f}s < {min_delay:.1f}s")
+        elif actual_delay > max_delay:
+            self.log.warning("Restart delayed beyond expected: %.1fs > %.1fs "
+                             "(may be acceptable depending on system load)",
+                             actual_delay, max_delay)
+        else:
+            self.log.info("SUCCESS: Restart delay within expected range [%.1fs, %.1fs]",
+                          min_delay, max_delay)
```
Do we really want to just log a warning if `actual_delay > max_delay`? For example what if the actual is like 500% more? I'd recommend making the max delay higher and then strictly enforcing that the actual delay is within bounds. This might eventually fail, but then we can revisit and increase the max again if needed. Otherwise, we might silently skip over an extremely high delay

tanabarr replied:
ok, done

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_advanced.yaml:
```diff
@@ -0,0 +1,26 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos0
```
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_advanced.yaml:
```diff
@@ -0,0 +1,26 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos1
```
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_disabled.py:
```diff
@@ -0,0 +1,152 @@
+
+        # Manually start the rank
+        self.log_step("Manually starting rank {test_rank}")
+        self.dmg.system_start(ranks=f"{test_rank}")
```
I think this should handle an int?
```suggestion
        self.dmg.system_start(ranks=test_rank)
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_disabled.py:
```diff
@@ -0,0 +1,152 @@
+        self.log_step("Manually restart ranks")
+
+        for rank in test_ranks:
+            self.dmg.system_start(ranks=f"{rank}")
```
```suggestion
            self.dmg.system_start(ranks=rank)
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_disabled.yaml:
```diff
@@ -0,0 +1,26 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos0
```
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/control/engine_auto_restart_disabled.yaml:
```diff
@@ -0,0 +1,26 @@
+      storage:
+        0:
+          class: ram
+          scm_mount: /mnt/daos1
```
```suggestion
      storage: auto
```

daltonbohning commented on src/tests/ftest/util/control_test_base.py:
```diff
@@ -46,3 +49,165 @@ def verify_dmg_storage_scan(self, verify_method):
+        self.dmg.system_exclude(ranks=[rank], rank_hosts=None)
+
+        # Wait for rank to self-terminate (should go to AdminExcluded state)
+        self.log_step(f"Waiting for rank {rank} to self-terminate")
+        time.sleep(2)
+
+        # Check if rank is adminexcluded
```
Similar to some other cases: how do we know 2s is enough? If there really is not a deterministic way to know at this code level, this is fine, but eventually this kind of thing needs to be revisited.

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_disabled.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_disabled.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos/job/PR-16279/33/display/redirect

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_advanced.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_disabled.py

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_advanced.yaml

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

tanabarr committed:
Update src/tests/ftest/control/engine_auto_restart_disabled.py

Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

mjmac commented on src/engine/drpc_ras.c:
Should `wait_for_resp` be `true` now, in order to ensure that the event processing doesn't race with the SIGKILL? I don't really see a downside...

tanabarr committed:
fail if delay > 200% of expected

Test-tag: hw,medium,dmg,control,engine_auto_restart pr
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

mjmac commented on src/control/server/instance_restart.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	// Can restart immediately
+	mgr.performRestart(ctx, rank, instance)
```
There is still a race here, I think. If multiple restart requests come in at the same time, there's a window where the manager goroutine and the AfterFunc goroutine could both wind up calling `performRestart()`.

If you rewrite this method to make the decision under the lock, I think you eliminate the race and also easily deal with floods of restart requests:
```go
      rank, instance := req.rank, req.instance

      mgr.mu.Lock()
      if last, ok := mgr.lastRestart[rank]; ok {
          if elapsed := time.Since(last); elapsed < mgr.getMinDelay() {
              // Fast debounce for subsequent requests inside of the delay window
              if _, pending := mgr.pendingRestart[rank]; pending {
                  mgr.mu.Unlock()
                  mgr.log.Debugf("rank %d already has a deferred restart pending; dropping",
                      rank)
                  return
              }

              // First restart request inside of the delay window claims it
              remaining := mgr.getMinDelay() - elapsed
              mgr.pendingRestart[rank] = time.AfterFunc(remaining, func() {
                  mgr.requestRestart(rank, instance)
              })
              mgr.mu.Unlock()
              mgr.log.Noticef("rank %d restart rate limited: will restart in %s",
                  rank, remaining.Round(time.Second))
              return
          }
      }
      
      // If this is the first restart or it's outside of the delay window, start the process (over)
      mgr.lastRestart[rank] = time.Now()
      delete(mgr.pendingRestart, rank)
      mgr.mu.Unlock()

      if err := waitForEngineStopped(ctx, []Engine{instance}); err != nil {
          mgr.log.Errorf("rank %d did not stop before restart: %s", rank, err)
          return
      }
      mgr.log.Noticef("restart manager is restarting rank %d", rank)
      instance.requestStart(ctx)
```

tanabarr replied:
would be acceptable to address this as a follow-on PR? or do you think it is a blocking issue?

kjacque replied:
I'd be fine with doing it as a follow-on, if only because I don't think a flood of near-simultaneous requests for the same node are likely.

tanabarr replied:
addressed in this change

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/45/execution/node/727/log

tanabarr committed:
try to solve NoneType iterate ftest failures

Test-tag: hw,medium,dmg,control,engine_auto_restart pr
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16279/46/execution/node/1364/log

daltonbohning reviewed (commented):
ftest LGTM

tanabarr committed:
return list of errors in reset_engine_restart_state (#18306)

Signed-off-by: Dalton Bohning <dalton.bohning@hpe.com>
Co-authored-by: Dalton Bohning <dalton.bohning@hpe.com>

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/48/testReport/

tanabarr committed:
fix pool sizes

Test-tag: hw,medium,dmg,control,engine_auto_restart pr
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

kjacque reviewed (dismissed):
Only looked at Go and C. I think the issues can be addressed in a follow-on, if @mjmac agrees.

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16279/49/testReport/

tanabarr committed:
refactor to simplify locking and remove race

Test-tag: hw,medium,dmg,control,engine_auto_restart pr
Signed-off-by: Tom Nabarro <thomas.nabarro@hpe.com>

mjmac reviewed (approved):
LGTM, thanks for reworking it

kjacque commented on src/control/server/instance_restart.go:
```diff
@@ -52,108 +52,50 @@ func (mgr *engineRestartManager) getMinDelay() time.Duration {
+		if elapsed := time.Since(last); elapsed < mgr.getMinDelay() {
+			// Fast debounce for subsequent requests inside of the delay window
+			if _, pending := mgr.pendingRestart[rank]; pending {
+				mgr.mu.Unlock()
```
suggestion - I'd be inclined to wrap this section happening under lock in a helper function so you can use `defer`.

tanabarr replied:
will rework when I revisit this code

kjacque reviewed (approved):
Suggestion is nonblocking. Thanks!

daltonbohning reviewed (commented):
ftest LGTM
